model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapStorage.java

```diff
@@ -149,79 +153,77 @@ public Crud(Class<V> entityClass, Function<String, Path> dataDirectoryFunc, Func
               .ifPresent(key -> this.fieldPredicates.replace(key, (builder, op, params) -> builder));
         }
 
-        protected Path getPathForSanitizedId(Path sanitizedIdPath) {
-            final Path dataDirectory = getDataDirectory();
-            final Path dataDirectoryWithChildren = dataDirectory.resolve(sanitizedIdPath).getParent();
-
-            if (! Files.isDirectory(dataDirectoryWithChildren)) {
-                try {
-                    Files.createDirectories(dataDirectoryWithChildren);
-                } catch (IOException ex) {
-                    throw new IllegalStateException("Directory does not exist and cannot be created: " + dataDirectory, ex);
+        protected Path getPathForEscapedId(String[] escapedIdPathArray) {
+            Path parentDirectory = getDataDirectory();
+            Path targetPath = parentDirectory;
+            for (String path : escapedIdPathArray) {
+                targetPath = targetPath.resolve(path).normalize();
+                if (! targetPath.getParent().equals(parentDirectory)) {
+                    LOG.warnf("Path traversal detected: %s", Arrays.toString(escapedIdPathArray));
+                    return null;
                 }
+                parentDirectory = targetPath;
             }
-            return dataDirectoryWithChildren.resolve(sanitizedIdPath.getFileName() + FILE_SUFFIX);
+
+            return targetPath.resolveSibling(targetPath.getFileName() + FILE_SUFFIX);
         }
 
-        protected Path getPathForSanitizedId(String sanitizedId) {
-            if (sanitizedId == null) {
-                throw new IllegalStateException("Invalid ID to sanitize");
+        protected Path getPathForEscapedId(String escapedId) {
+            if (escapedId == null) {
+                throw new IllegalStateException("Invalid ID to escape");
             }
 
-            return getPathForSanitizedId(Path.of(sanitizedId));
+            String[] escapedIdArray = ID_COMPONENT_SEPARATOR_PATTERN.split(escapedId);
+            return getPathForEscapedId(escapedIdArray);
         }
 
-        protected String sanitizeId(String id) {
-            Objects.requireNonNull(id, "ID must be non-null");
+        // Percent sign + Unix (/) and https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file reserved characters
```

Comment:
The reserved chars listed here are only a subset of those reserved characters listed at the given URL. 

* Integer value zero, sometimes referred to as the ASCII NUL character.
* Characters whose integer representations are in the range from 1 through 31

I assume that the creation of such a file will fail, and there will be no harm done. As those file names would also be strange on Linux systems, I'd suggest to treat those characters as illegal and not even escape those. 

The docs also list LPT1 and other file names which shouldn't be used. I gave it a try on a recent Windows machine:  I was able to create those files with ported *nix tools on the file system, but wasn't able to retrieve the files with *OS/*indows tools, so the results are kind of mixed. I am not sure if harm can be done here, so I'd suggest to avoid those names on Windows file systems.

As those items are not necessarily related to the path traversal issue, feel free to consider them for a future issue.

Reply:
> I assume that the creation of such a file will fail, and there will be no harm done

That is correct, and thus is not a concern too. I do not see the need of fixing those, since even in windows, they can be used inside a .zip archive which is one of the future paths for the file store.